### DISCUSSION TITLE
Package and distribute TDEX Daemon UI

### DISCUSSION THREAD

**Discussion by tiero** [TDex-network/tdex-daemon#490] [Ideas]
## Overview

The [TDEX Daemon UI](https://tdex-network.github.io/tdex-daemon-ui/) is a grpc-web client that connects to a `tdexd` instance and  in order to securely encrypts the connection (especially if run on a remote instance) we use TLS self-signed certificate, which [MUST be installed in the end-user system](https://www.entrust.com/knowledgebase/ssl/how-to-install-ssl-certificate-on-a-mac-osx-107-server-and-up) before he's able to securely use TLS termination from his browser (and not have browser complain about certificate not trusted, preventing API calls for CORS policies)

## Where we want to ship

We want to be able to ship the app to user in many ways, desktop app (compiled with Tauri?), web app on github pages and on the node-in-a-box projects (ie. Umbrel)  

## Solution

For the web app solution we can only teach users how to do this (easier path on us, more clunky pairing UX, especially if we serve in node-in-a-box project ie. Umbrel), but we can improve this shipping an extra binary for major platforms (MacOS, Linux and Windows) that works as gRPC-Web to gRPC proxy, which will listen for un-encrypted (and un-autheticated also?) calls from localhost and proxy them using a normal gRPC client (ie. written in golang since we have the tdex CLI?)

This same approach is used by Lightning Terminal, which does the following:

1) ships his UI as go binary where it [embeds the HTML,CSS,JS bundle using go embed directive ](https://github.com/lightninglabs/lightning-terminal/blob/master/terminal.go#L661) and serves them with a plain http listener
2) the binary also exposes a grpc-web listener on the same port
3) the UI calls the localhost grpc-web server unencrypted, which [proxies](https://github.com/lightninglabs/lightning-terminal/blob/master/rpc_proxy.go#L97) the call using a native go gRPC client to the actual gRPC server (remote or integrated) using TLS encrypted (and authenticated)

### Desktop App with Tauri

We can make it this a bit easier (ie. without having the user to install a go binary manually) which is shipping the UI as dektop app with [Tauri](https://tauri.studio) which is installable as "one-click" and package the go binary as a [sidecar](https://tauri.studio/en/docs/usage/guides/bundler/sidecar/)  and use the internal [bridge system](https://tauri.studio/en/docs/usage/patterns/bridge/#configuration) to save in a well-known datadir the TLS so the go binary can access it (given in the UI pasted via `tdexdconnect` URL)

### Browser app in Umbrel 

In umbrel (and other solutions not docker based) where we have more flexibility, we will run a docker-compose with three services ideally: `tdexd`, `rpcproxy` and `tdex-daemon-ui` manually and provide the right path for the proxy how to reach tdexd's datadir with the certificate/macaroon

**Comment by altafan** [comment]
>For the web app solution we can only teach users how to do this (easier path on us, more clunky pairing UX, especially if we serve in node-in-a-box project ie. Umbrel), but we can improve this shipping an extra binary for major platforms (MacOS, Linux and Windows) that works as gRPC-Web to gRPC proxy, which will listen for un-encrypted (and un-autheticated also?) calls from localhost and proxy them using a normal gRPC client (ie. written in golang since we have the tdex CLI?)

So the web app hosted on github pages would require the user to run a proxy on its local machine right? This proxy could be configured via a tdexdconnect URL string, or by providing the path to the macaroon and TLS cert files if needed.

Anytime the user wanted to access the web-app UI, he should remember to start the proxy, shouldn't it?

**Reply by tiero** [reply]
We assume that who uses the app from Github Pages deployment, will paste the tdexconnect URL and the frontend via JavaScript will download the TLS certificate as file, so the user must double click and install in his system.

Also in the desktop app, the user paste his tdexconnect URL, but does not need to store it & install, since the desktop app will store automatically in his own datadir  using the filesystem API (ie. Tauri bridge). When run in desktop, the go proxy is packaged as sidecar, that will source from the well know datadir the TLS (and macaroon) certificate file to be passed to the gRPC client the proxy uses to call the `tdexd`

**Reply by tiero** [reply]
What you propose (web app on github pages and proxy on loclahost) is doable, the githiub pages will hardcode localhost, just not convenient imho: since user is about to download something (ie. proxy) let him download the whole web packaged as webview in a desktop app, so UX is more: download and click and you there.
